Fix languages lookup assert

CrystalP opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

LangCodeExpander::Lookup could trigger the assert in StringToLongCode() with BCP 47 language tags due to not checking length correctly prior to the call as the code before #27229 used to. Matching unit test added.

The appearance change of deprecated ISO 639-1 language codes in the languages list sparked some debate and bike shedding. Since it wasn't the purpose of the previous PRs or this one, the appearance is reverted to that prior to the PRs (=identical looking languages lines in the list), to be modified in followup PRs.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

assert triggered by ISA

debate around the appearance of deprecated languages.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Added unit tests.

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

Hopefully clearer language list.

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [X] I have added tests to cover my change
- [X] All new and existing tests passed

CrystalP committed:
[i18n] Fix lookup of BCP 47 language tags

CrystalP commented:
If you're OK with that then I should translated messages in strings.po instead of hardcoded text "Deprecated" "Non-preferred"

CastagnaIT commented on xbmc/utils/i18n/Iso639_1.cpp:
```diff
@@ -62,9 +62,13 @@ bool CIso639_1::ListLanguages(std::map<std::string, std::string>& langMap)
+                         {
+                           std::string c = LongCodeToString(e.code);
+                           return std::make_pair(c, std::string{e.name} + " \"" + c +
+                                                        "\" (Deprecated)");
```
Even if they are removed in the future, a user can always add them back with XML advanced settings

CastagnaIT replied:
im not talking about to remove from the list the "deprecated" and "no-preferred"

im talking about GUI view:
<img width="832" height="258" alt="immagine" src="https://github.com/user-attachments/assets/dd3b36db-ac7f-4ff5-813c-78265edb306d" />
the `(Deprecated)` text has no real meaning for end-user and this text must be removed from GUI
If a subtitle or audio track uses this code, it makes no sense for the user to know that it is deprecated; it is only an internal matter for devs
and for consistence same thing `non-preferred` text from similar 6392 method must be removed (not whole code it can stay for me...)

and from the audio or subtitle list while in playback are indistinguishable
<img width="706" height="121" alt="immagine" src="https://github.com/user-attachments/assets/fa00743c-b7b9-486b-9b6e-bd9529c8eea2" />
I would have made the distinction with the language code here as well for consistency, this makes it clear exactly what it is

CrystalP replied:
I'm not against that type of change, but it's beyond the scope of this and past PR. It's taking too much time.
I spent way too much time on this language refresh as it is and it's preventing me from tackling BCP 47 and support of movies/tv shows original language that I wanted to get into v22, the reason for touching the subject.

I'll put things back how they were before my PRs (= identical looking language lines) and will let you do what you want with the GUI in a followup PR.
As written before, I don't know how languages are used throughout Kodi and wasn't attempting to make improvements to the user experience.

Some of what you want to do seems to have its place in skin rather than forcing formats with core by the way and core should probably just provide the information as separate fields that the skin can format however it wants.

CastagnaIT commented:
I think the alternative fix will work just as well, despite the fact that i dislike asserts
I'll do a confirmation test tomorrow

garbear commented:
> despite the fact that i dislike asserts

Asserts are good, but when one is triggered by distant code, it means it should have been a safety check instead. That assert that triggered should be replaced with proper handling for other error cases from distant code.

CrystalP committed:
[i18n] Doxygen grammar fixes

CrystalP committed:
[i18n] Revert previous PRs changes to english name in languages lists

CrystalP commented:
Fixed clang format, fixed unit test comment, fixed doxygen comment, stop listing the non-preferred 639-2 B codes.

If the deprecated 639-1 codes can be omitted from the languages list I'll apply that change instead of attempting to disambiguate the codes in the listing.

The assert was triggered by a direct caller, not some distant code. There is probably a better way to code this, I was only getting rid of the old-style #define MAKECODE. The function is private to ISO-code handling functions and not meant for callers outside of LangCodeExpander, gateway for codes and language names conversions and lookups.

CastagnaIT reviewed (changes requested):
please address better GUIs

CastagnaIT reviewed (approved):
> Not the scope of the PR and it's taking too long so I revert to how things were before any of my changes. You do it however you please.

these changes has been introduced from your previous rework prs
so i expect these aspects to be fixed in time for the release
you come across as a bit arrogant, i never asked to rewrite all the code if just deleting the not user-friendly texts, by keeping language code to disambiguate the two languages

but if you prefer to revert to the original behavior, that's fine with me anyway
